--- /home/benjaminm/code/oops-bundle/ecsaber/saber/saber-import.cmake.in.tmp	2023-11-10 09:38:12.736232448 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/saber-import.cmake.in	2023-11-07 15:34:43.597507332 +0100
@@ -3,10 +3,6 @@
 
 include(CMakeFindDependencyMacro)
 
-if(@jedicmake_FOUND@ AND NOT jedicmake_FOUND)
-    find_dependency(jedicmake REQUIRED)
-endif()
-
 if(@OPENMP@)
     find_dependency(OpenMP REQUIRED COMPONENTS Fortran)
 endif()
@@ -45,6 +41,11 @@
     set(saber_oops_FOUND True) #COMPONENT 'oops'
 endif()
 
+if(@oops-patch_FOUND@) # oops-patch_FOUND
+    find_dependency(oops-patch)
+    set(saber_oops-patch_FOUND True) #COMPONENT 'oops-patch'
+endif()
+
 if(@gsibec_FOUND@) # gsibec_FOUND
     find_dependency(gsibec)
     set(saber_gsibec_FOUND True) #COMPONENT 'gsibec'
