--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/spectralb/CovarianceStatisticsUtils.cc.tmp	2023-11-10 16:12:30.128284717 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/spectralb/CovarianceStatisticsUtils.cc	2023-11-08 13:53:36.697553011 +0100
@@ -362,7 +362,7 @@
 
   // read from header file on root PE.
   std::size_t root = 0;
-  if (eckit::mpi::comm().rank() == root) {
+  if (oops::mpi::world().rank() == root) {
      util::atlasArrayInquire(ncfilepath,
                              dimNames,
                              dimSizes,
@@ -378,7 +378,7 @@
   auto specvertview = atlas::array::make_view<double, 3>(spectralVertCov);
   specvertview.assign(0.0);
 
-  if (eckit::mpi::comm().rank() == root) {
+  if (oops::mpi::world().rank() == root) {
     auto it = std::find(variableNames.begin(), variableNames.end(), filevar);
     std::size_t i = std::distance(variableNames.begin(), it);
     std::vector<atlas::idx_t> dimSizesForVar;
@@ -394,9 +394,9 @@
                              netcdfVarIDs[i],
                              specvertview);
   }
-  util::scatter<double>(eckit::mpi::comm(), root, specvertview);
+  util::scatter<double>(oops::mpi::world(), root, specvertview);
 
-  if (eckit::mpi::comm().rank() == root) {
+  if (oops::mpi::world().rank() == root) {
     const int retval = nc_close(netcdfGeneralIDs[0]);
     if (retval) ERR(retval);
   }
