--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/spectralb/SpectralToGauss.h.tmp	2023-11-10 09:38:10.172230970 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/spectralb/SpectralToGauss.h	2023-11-08 13:53:36.413552653 +0100
@@ -63,14 +63,12 @@
 
   atlas::FieldSet generateInnerFieldSet(const oops::GeometryData & innerGeometryData,
                                         const oops::patch::Variables & innerVars) const override
-    {return util::createSmoothFieldSet(innerGeometryData.comm(),
-                                       innerGeometryData.functionSpace(),
+    {return util::createSmoothFieldSet(innerGeometryData.functionSpace(),
                                        innerVars);}
 
   atlas::FieldSet generateOuterFieldSet(const oops::GeometryData & outerGeometryData,
                                         const oops::patch::Variables & outerVars) const override
-    {return util::createSmoothFieldSet(outerGeometryData.comm(),
-                                       outerGeometryData.functionSpace(),
+    {return util::createSmoothFieldSet(outerGeometryData.functionSpace(),
                                        outerVars);}
 
  private:
