--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/interpolation/GaussToCS.h.tmp	2023-11-10 16:12:27.632288018 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/interpolation/GaussToCS.h	2023-11-08 13:53:36.101552261 +0100
@@ -80,14 +80,12 @@
 
   atlas::FieldSet generateInnerFieldSet(const oops::GeometryData & innerGeometryData,
                                         const oops::patch::Variables & innerVars) const override
-    {return util::createSmoothFieldSet(innerGeometryData.comm(),
-                                       innerGeometryData.functionSpace(),
+    {return util::createSmoothFieldSet(innerGeometryData.functionSpace(),
                                        innerVars);}
 
   atlas::FieldSet generateOuterFieldSet(const oops::GeometryData & outerGeometryData,
                                         const oops::patch::Variables & outerVars) const override
-    {return util::createSmoothFieldSet(outerGeometryData.comm(),
-                                       outerGeometryData.functionSpace(),
+    {return util::createSmoothFieldSet(outerGeometryData.functionSpace(),
                                        outerVars);}
 
  private:
