--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/blocks/SaberBlockChainBase.h.tmp	2023-12-03 10:02:37.570068388 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/blocks/SaberBlockChainBase.h	2023-12-03 10:00:15.949727416 +0100
@@ -21,6 +21,7 @@
 }
 
 namespace oops {
+  class FieldSet3D;
   class FieldSet4D;
   template <class MODEL> class Geometry;
   namespace patch{
