--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/blocks/SaberCentralBlockBase.h.tmp.bak	2023-12-18 08:28:34.168183422 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/blocks/SaberCentralBlockBase.h	2023-12-15 16:39:01.871508180 +0100
@@ -30,6 +30,7 @@
 #include "oops/util/Printable.h"
 
 #include "saber/blocks/SaberBlockParametersBase.h"
+#include "saber/oops/ECUtilities.h"
 
 // Forward declaration
 namespace oops {
@@ -226,12 +227,17 @@
   // Read fieldsets as increments
   std::vector<oops::FieldSet3D> fsetVec;
   for (const auto & input : this->getReadConfs()) {
+    // Create variables
+    oops::Variables<MODEL> varsT(templatedVarsConf(vars));
+
     // Create increment
-    oops::Increment<MODEL> dx(geom, vars, validTime_);
+    oops::Increment<MODEL> dx(geom, varsT, validTime_);
     dx.read(input.second);
     oops::Log::test() << "Norm of input parameter " << input.first
                       << ": " << dx.norm() << std::endl;
-    fsetVec.push_back(dx.fieldSet());
+    oops::FieldSet3D fset(validTime_, eckit::mpi::comm());
+    fset.deepCopy(dx.increment().fieldSet());
+    fsetVec.push_back(fset);
     fsetVec.back().name() = input.first;
   }
   this->setReadFields(fsetVec);
@@ -250,13 +256,16 @@
   std::vector<std::pair<eckit::LocalConfiguration, oops::FieldSet3D>> outputs
     = this->fieldsToWrite();
 
+  // Create variables
+  oops::Variables<MODEL> varsT(templatedVarsConf(vars));
+
   // Create increment
-  oops::Increment<MODEL> dx(geom, vars, validTime_);
+  oops::Increment<MODEL> dx(geom, varsT, validTime_);
 
   // Write fieldsets as increments
   for (const auto & output : outputs) {
-    dx.fieldSet().deepCopy(output.second);
-    dx.synchronizeFields();
+    dx.increment().fieldSet() = output.second.fieldSet();
+    dx.increment().synchronizeFields();
     oops::Log::test() << "Norm of output parameter " << output.second.name()
                       << ": " << dx.norm() << std::endl;
     dx.write(output.first);
