--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/oops/Utilities.cc.tmp	2023-11-10 16:12:30.592284103 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/oops/Utilities.cc	2023-11-08 13:39:01.135441521 +0100
@@ -15,7 +15,7 @@
 // -----------------------------------------------------------------------------
 
 oops::patch::Variables getActiveVars(const SaberBlockParametersBase & params,
-                              const oops::patch::Variables & defaultVars) {
+                                     const oops::patch::Variables & defaultVars) {
   oops::Log::trace() << "getActiveVars starting" << std::endl;
   oops::patch::Variables activeVars;
   if (params.mandatoryActiveVars().size() == 0) {
@@ -40,6 +40,22 @@
 }
 
 // -----------------------------------------------------------------------------
+
+void setMember(eckit::LocalConfiguration & conf,
+               const int & member) {
+  oops::Log::trace() << "setMember starting" << std::endl;
+
+  if (conf.has("member pattern")) {
+    std::string memberPattern = conf.getString("member pattern");
+    util::seekAndReplace(conf, memberPattern, std::to_string(member));
+  } else {
+    conf.set("member", member);
+  }
+
+  oops::Log::trace() << "setMember done" << std::endl;
+}
+
+// -----------------------------------------------------------------------------
 
 void setMPI(eckit::LocalConfiguration & conf,
             const int & mpi) {
