--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/oops/Utilities.cc.tmp.bak	2023-12-19 10:47:24.441034147 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/oops/Utilities.cc	2023-12-19 10:38:35.469135949 +0100
@@ -42,6 +42,22 @@
 
 // -----------------------------------------------------------------------------
 
+void setMember(eckit::LocalConfiguration & conf,
+               const int & member) {
+  oops::Log::trace() << "setMember starting" << std::endl;
+
+  if (conf.has("member pattern")) {
+    std::string memberPattern = conf.getString("member pattern");
+    util::seekAndReplace(conf, memberPattern, std::to_string(member));
+  } else {
+    conf.set("member", member);
+  }
+
+  oops::Log::trace() << "setMember done" << std::endl;
+}
+
+// -----------------------------------------------------------------------------
+
 void setMPI(eckit::LocalConfiguration & conf,
             const int & mpi) {
   oops::Log::trace() << "setMPI starting" << std::endl;
