--- /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/mpi/mpi.cc.tmp	2023-11-09 17:31:28.137939397 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/mpi/mpi.cc	2023-11-07 15:34:43.597507332 +0100
@@ -189,7 +189,7 @@
 }
 
 // ------------------------------------------------------------------------------------------------
-void broadcastBool(const eckit::mpi::Comm & comm, bool & boolVar, const size_t root) {
+void broadcastBool(const eckit::mpi::Comm & comm, bool & boolVar, const int root) {
     // Send bool as int since eckit MPI broadcast doesn't accept bool
     int tempInt;
     if (comm.rank() == root) {
@@ -201,7 +201,7 @@
     }
 }
 
-void broadcastString(const eckit::mpi::Comm & comm, std::string & stringVar, const size_t root) {
+void broadcastString(const eckit::mpi::Comm & comm, std::string & stringVar, const int root) {
     std::vector<char> buffer;
     if (comm.rank() == root) {
         // Send string as vector of char since eckit MPI broadcast doesn't accept string
