--- /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/mpi/mpi.cc.tmp	2023-11-16 14:47:14.423971668 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/mpi/mpi.cc	2023-11-16 14:45:47.900081646 +0100
@@ -154,9 +154,15 @@
 void allGatherv(const eckit::mpi::Comm & comm, std::vector<util::DateTime> &x) {
     size_t globalSize = x.size();
     comm.allReduceInPlace(globalSize, eckit::mpi::sum());
-    std::vector<util::DateTime> globalX(globalSize);
-    oops::mpi::allGathervUsingSerialize(comm, x.begin(), x.end(), globalX.begin());
-    x = std::move(globalX);
+    std::vector<std::string> xStr;
+    for (const auto & item : x) {
+      xStr.push_back(item.toString());
+    }
+    oops::mpi::allGatherv(comm, xStr);
+    x.clear();
+    for (const auto & item : xStr) {
+      x.push_back(util::DateTime(item));
+    }
 }
 
 // ------------------------------------------------------------------------------------------------
