--- /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/CMakeLists.txt.tmp.bak	2024-01-25 15:54:44.638476054 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/saber/src/saber/CMakeLists.txt	2024-01-25 15:59:53.646930867 +0100
@@ -29,6 +29,7 @@
 target_link_libraries( ${PROJECT_NAME} PUBLIC fckit )
 target_link_libraries( ${PROJECT_NAME} PUBLIC atlas_f )
 target_link_libraries( ${PROJECT_NAME} PUBLIC ${oops_LIBRARIES} )
+target_link_libraries( ${PROJECT_NAME} PUBLIC oops-patch )
 
 if( gsibec_FOUND )
     target_link_libraries( ${PROJECT_NAME} PUBLIC gsibec )
