--- /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/base/FieldSet4D.h.tmp.bak	2023-12-15 16:07:00.098690129 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/base/FieldSet4D.h	2023-12-15 14:52:37.165851881 +0100
@@ -12,8 +12,8 @@
 
 #include "oops/base/FieldSet3D.h"
 #include "oops/base/FieldSets.h"
-#include "oops/base/Increment4D.h"
-#include "oops/base/State4D.h"
+#include "oops/assimilation/Increment4D.h"
+#include "oops/assimilation/State4D.h"
 
 namespace oops {
 
@@ -67,11 +67,11 @@
 
 template<typename MODEL>
 FieldSet4D::FieldSet4D(const State4D<MODEL> & state4d)
-  : FieldSets(state4d.times(), state4d.commTime(), {0}, oops::mpi::myself()) {
-  for (size_t jj = 0; jj < state4d.size(); ++jj) {
+  : FieldSets(state4d.times(), eckit::mpi::self(), {0}, eckit::mpi::self()) {
+  for (size_t jj = 0; jj < state4d.times().size(); ++jj) {
     this->dataset().emplace_back(new FieldSet3D(state4d[jj].validTime(),
-                                                state4d.geometry().getComm()));
-    this->dataset()[jj]->shallowCopy(state4d[jj].fieldSet());
+                                                eckit::mpi::comm()));
+    this->dataset()[jj]->shallowCopy(state4d[jj].state().fieldSet());
   }
 }
 
@@ -79,11 +79,11 @@
 
 template<typename MODEL>
 FieldSet4D::FieldSet4D(const Increment4D<MODEL> & inc4d)
-  : FieldSets(inc4d.times(), inc4d.commTime(), {0}, oops::mpi::myself()) {
-  for (size_t jj = 0; jj < inc4d.size(); ++jj) {
+  : FieldSets(inc4d.times(), eckit::mpi::self(), {0}, eckit::mpi::self()) {
+  for (size_t jj = 0; jj < inc4d.times().size(); ++jj) {
     this->dataset().emplace_back(new FieldSet3D(inc4d[jj].validTime(),
-                                                inc4d.geometry().getComm()));
-    this->dataset()[jj]->shallowCopy(inc4d[jj].fieldSet());
+                                                eckit::mpi::comm()));
+    this->dataset()[jj]->shallowCopy(inc4d[jj].increment().fieldSet());
   }
 }
 
