--- /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/base/ModelSpaceCovarianceParametersBase.h.tmp	2023-11-09 17:31:27.993939546 +0100
+++ /home/benjaminm/code/oops-bundle/ecsaber/oops-patch/src/oops/base/ModelSpaceCovarianceParametersBase.h	2023-11-07 15:34:43.597507332 +0100
@@ -11,7 +11,6 @@
 #include <string>
 #include <vector>
 
-#include "oops/interface/LinearVariableChange.h"
 #include "oops/util/parameters/OptionalParameter.h"
 #include "oops/util/parameters/Parameter.h"
 #include "oops/util/parameters/Parameters.h"
@@ -32,15 +31,12 @@
   /// a covariance model whose type is determined at runtime), but not others (e.g. in tests
   /// written with a particular model in mind). ModelSpaceCovarianceParametersWrapper will throw an
   /// exception if this parameter is not provided.
-  typedef typename LinearVariableChange<MODEL>::Parameters_ Parameters_;
   OptionalParameter<std::string> covarianceModel{"covariance model", this};
 
   OptionalParameter<size_t> randomizationSize{"randomization size", this};
   Parameter<bool> fullInverse{"full inverse", false, this};
   Parameter<int> fullInverseIterations{"full inverse iterations", 10, this};
   Parameter<double> fullInverseAccuracy{"full inverse accuracy", 1.0e-3, this};
-  OptionalParameter<Parameters_> variableChange{"linear variable change",
-    this};
 };
 
 }  // namespace oops
